reference-implementation/lib/readable-stream.js

```diff
@@ -484,3 +484,157 @@ function ShouldReadableStreamApplyBackpressure(stream) {
 
   return shouldApplyBackpressure;
 }
+
+function TeeReadableStream(stream, clone) {
+  const reader = stream.getReader();
+
+  let enqueue1, enqueue2, close1, close2, error1, error2;
+  let canceled1, cancelReason1, canceled2, cancelReason2;
+  const branch1 = new ReadableStream({
+    start(enqueue, close, error) {
+      [enqueue1, close1, error1] = [enqueue, close, error];
+    },
+    pull: readAndEnqueueInBoth,
+    cancel(reason) {
+      canceled1 = true;
+      cancelReason1 = reason;
+      maybeCancelSource();
+    }
+  });
+  const branch2 = new ReadableStream({
+    start(enqueue, close, error) {
+      [enqueue2, close2, error2] = [enqueue, close, error];
+    },
+    pull: readAndEnqueueInBoth,
+    cancel(reason) {
+      canceled2 = true;
+      cancelReason2 = reason;
+      maybeCancelSource();
+    }
+  });
+
+  return [branch1, branch2];
+
+  function readAndEnqueueInBoth() {
+    reader.read().then(
+      ({ value, done }) => {
+        if (done) {
+          close1();
+          close2();
+          return;
+        }
+
+        if (clone) {
+          enqueue1(StructuredClone(value));
+          enqueue2(StructuredClone(value));
+        } else {
+          enqueue1(value);
+          enqueue2(value);
+        }
+      },
+      e => {
+        error1(e);
+        error2(e);
+      }
+    );
+  }
+
+  function maybeCancelSource() {
+    if (canceled1 && canceled2) {
+      reader.cancel([cancelReason1, cancelReason2]);
```

Comment:
Fixed. It turns out to be a bit more complicated since we need to return a promise even to the first person to cancel, not just the second.

Unless we think the first should fulfill ASAP and the second should signal? I will mention below so it doesn't get lost in this old diff.